### PR TITLE
Port Qunit tests to cdp

### PR DESCRIPTION
**Pull request by larskarlitski**
Port QUnit tests to chrome debug protocol, reusing @martinpitt's work for the integration tests.

Marked as WIP, because:
- [x] add chromium to unit-test container for semaphore (#8349)
- [x] figure out how to subscribe to resource loading errors over CDP (see TODO in tap-cdp.py)
- [ ] remove workarounds we have for phantom behaving weirdly (if possible)
- [x] land #8350

followup cleanups:
- [ ] clean up injected scripts: move them from cdp.py to testlib.py and drop bool parameter

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -254,11 +254,7 @@ TEST_EXTENSIONS = .html .sh
 SH_LOG_DRIVER = $(top_srcdir)/tools/tap-driver
 HTML_LOG_DRIVER = $(top_srcdir)/tools/tap-driver
 HTML_TEST_SERVER = $(builddir)/test-server
-if WITH_PHANTOMJS
-HTML_LOG_COMPILER = $(PHANTOMJS) $(top_srcdir)/tools/tap-phantom --strip=$(top_srcdir)/ -- $(HTML_TEST_SERVER)
-else
-HTML_LOG_DRIVER_FLAGS = --missing=phantomjs
-endif
+HTML_LOG_COMPILER = $(top_srcdir)/test/common/tap-cdp --strip=$(top_srcdir)/ -- $(HTML_TEST_SERVER)
```
Comment (martinpitt) [inline Makefile.am]: You still need the `if`fery, as our test VM image build environments (mock, pbuilder) don't have chromium installed (nor phantomjs).

**Comment by martinpitt** [comment]
@larskarlitski: Nice, thanks for working on this! I didn't review yet, but wrt. your questions about resource errors: They appear in [Log.entryAdded](https://chromedevtools.github.io/devtools-protocol/1-3/Log/#event-entryAdded), like [here](http://209.132.184.41/logs/pull-8069-20171208-065252-ec5d7ceb-verify-rhel-7/log.html#80):
```
CDP: {"source":"network","level":"error","text":"Failed to load resource: the server responded with a status of 401 (Authentication failed)","timestamp":1512717056493.03,"url":"http://127.0.0.2:9191/cockpit/login","networkRequestId":"26842.8"}
```
(of course this instance is an "expected" one). This is printed by the handler in test/common/cdp-driver.js.

Another promising API is [Network.loadingFailed](https://chromedevtools.github.io/devtools-protocol/1-2/Network/#event-loadingFailed), but I didn't experiment with that yet.

File: configure.ac [hunk 2]
```diff
@@ -269,12 +269,6 @@ else
   AC_MSG_ERROR(no. Please install intltool package)
 fi
 
-# phantomjs
-
-AC_PATH_PROG([PHANTOMJS], [phantomjs], [], [$srcdir/node_modules/.bin$PATH_SEPARATOR$PATH])
-AM_CONDITIONAL([WITH_PHANTOMJS], [test -n "$PHANTOMJS"])
-AC_SUBST([PHANTOMJS])
```
Comment (martinpitt) [inline configure.ac]: This needs to be replaced with logic to search for chromium, similar to [browser_path() in testlib.py](https://github.com/cockpit-project/cockpit/blob/master/test/common/testlib.py#L1265)

Comment (larskarlitski) [reply configure.ac]: Indeed, thanks. Fixed this first to make the environments without chromium installed work.

**Comment by martinpitt** [comment]
Thanks @larskarlitski ! I added a TODO item to add chromium to the unit-test container for semaphore.

**Comment by martinpitt** [comment]
Why do you need to split out the `CDP` class, OOI? Can't you just import the whole testlib from the tap driver?

**Comment by larskarlitski** [comment]
> Why do you need to split out the CDP class, OOI? Can't you just import the whole testlib from the tap driver?

To be honest, I can't think of the original reason anymore. I still think it's a bit cleaner, because the unit testing has no other common code with the integration tests. If we were to split out the integration tests at some point, this would make it easier. I don't feel strongly about it though and don't mind changing it if you prefer.

**Comment by martinpitt** [comment]
@larskarlitski : No, it's fine for stylistic reasons, too. I just wondered if I was missing something.

**Comment by martinpitt** [comment]
@larskarlitski : In PR #8349 I added chromium to the unit-tests container for semaphore, so you can start running tests there.

**Comment by larskarlitski** [comment]
Fixed the `jsquote` issue (I moved it into cdp.py, but testlib.py uses it in other places as well) and separated the refactor into its own pull request at #8350.

File: bots/images/scripts/fedora-26.setup [hunk 3]
```diff
@@ -103,7 +103,6 @@ dnf $DNF_OPTS -y upgrade
 dnf $DNF_OPTS -y install $TEST_PACKAGES $COCKPIT_DEPS $IPA_CLIENT_PACKAGES $ABRT_PACKAGES $AVOCADO_TEST_DEPS
 
 # for avocado/fedora integration test
-npm -g install phantomjs-prebuilt
```
Comment (martinpitt) [inline bots/images/scripts/fedora-26.setup]: I'd suggest to leave the removal of phantomjs from the images to a follow-up PR, to avoid having to rebuild/land images in this PR.

File: configure.ac [hunk 4]
```diff
@@ -269,11 +269,14 @@ else
   AC_MSG_ERROR(no. Please install intltool package)
 fi
 
-# phantomjs
+# chromium - same as test/common/cdp.py
 
-AC_PATH_PROG([PHANTOMJS], [phantomjs], [], [$srcdir/node_modules/.bin$PATH_SEPARATOR$PATH])
-AM_CONDITIONAL([WITH_PHANTOMJS], [test -n "$PHANTOMJS"])
-AC_SUBST([PHANTOMJS])
+AC_PATH_PROGS([CHROMIUM], [chromium-browsr chrome], [], [$srcdir/node_modules/chromium/lib/chromium/chrome-linux/chrome$PATH_SEPARATOR$PATH])
```
Comment (martinpitt) [inline configure.ac]: "browsr" typo

File: test/common/tap-cdp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+success = True
+
+server = subprocess.Popen([opts.server],
+                          stdout=subprocess.PIPE,
+                          stdin=subprocess.PIPE,
+                          close_fds=True)
+
+cdp = cdp.CDP("C.utf8")
+cdp.invoke("Page.navigate", url='file:///home/lars/src/cockpit/dist/docker/test-docker.html')
```
Comment (martinpitt) [inline test/common/tap-cdp]: I suppose this should be `opts.test` eventually? (I. e. this won't work on our CI like that)

File: test/common/tap-cdp [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+while True:
+    # TODO don't access private field
+    line = cdp._driver.stderr.readline()
+
+    _, message = driver_line_re.match(line).groups()
```
Comment (martinpitt) [inline test/common/tap-cdp]: That's a bit ugly IMHO -- `.group(2)` instead?

File: test/common/tap-cdp [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    elif message == 'cockpittest-tap-error':
+        sys.exit(1)
+    elif message == 'cockpittest-tap-expect-resource-error':
+        # TODO catch resource errors (and stop catching them here)
+        pass
+
+    if tap_line_re.match(message):
+        print(message)
+    else:
+        print(message, file=sys.stderr)
```
Comment (martinpitt) [inline test/common/tap-cdp]: supposedly this should set `success = False` here

File: bots/images/scripts/fedora-26.setup (hunk 3 above)
Comment (larskarlitski) [reply bots/images/scripts/fedora-26.setup]: Makes sense.

File: configure.ac (hunk 4 above)
Comment (larskarlitski) [reply configure.ac]: I think we should rename the package! ;)

File: test/common/tap-cdp (hunk 5 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Oooops, sorry. I used this for testing and forgot to remove it.

File: test/common/tap-cdp (hunk 6 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Yeah that's much better, thanks. I also moved it into `cdp.py` to remove that one TODO.

File: test/common/tap-cdp (hunk 7 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Yes, thanks.

File: test/common/cdp.py [hunk 8]
```diff
@@ -15,6 +15,8 @@
 
 TEST_DIR = os.path.normpath(os.path.dirname(os.path.realpath(os.path.join(__file__, ".."))))
 
+driver_line_re = re.compile(r'^> ([a-z]+): (.*)\n$')
```
Comment (martinpitt) [inline test/common/cdp.py]: shouldn't this be `console_line_re` rather?

File: test/common/cdp.py [hunk 9]
```diff
@@ -228,3 +230,10 @@ def set_frame(self, frame):
         self.cur_frame = frame
         if self.trace:
             print("-> switch to frame %s" % frame)
+
+    def read_console_line(self):
+        if not self.valid:
+            return
+
+        line = self._driver.stderr.readline()
+        return driver_line_re.match(line).group(2)
```
Comment (martinpitt) [inline test/common/cdp.py]: This looks a bit fiddly. Also it cannot work as the `self._driver` process does not redirect its stderr. This is deliberate, as we want the user to see these messages from the driver.

If this is supposed to grab the first (or last) console message, then use `self.get_js_log()[0]` (or `[-1]` for the last one).

File: test/common/cdp.py (hunk 8 above)
Comment (martinpitt) [reply test/common/cdp.py]: See below, this can just be dropped.

File: test/common/cdp.py (hunk 9 above)
Comment (larskarlitski) [reply test/common/cdp.py]: > This looks a bit fiddly. Also it cannot work as the self._driver process does not redirect its stderr. This is deliberate, as we want the user to see these messages from the driver.

It does as of #8350. We need this for the unit tests, because the tap tests output control messages such as `cockpittest-tap-done`, which we need to catch in `tap-cdp`.

I see your point for wanting to see these messages for the integration tests, though. We could either add an `stderr` parameter to the cdp class or find a different way of communicating status from the qunit tests.

**Comment by martinpitt** [comment]
> It does as of #8350

Argh! I didn't see that, and the commit message didn't mention that either. Can you please revert this? What's wrong with grabbing the console log messages from `CDP.get_js_log()`?

**Comment by larskarlitski** [comment]
Rebased. Now that `tap-cdp` cannot block on `cdp.stderr` anymore, we need to poll it via `get_js_log` (which I moved to the cdp class for this reason).

File: test/common/cdp-driver.js [hunk 10]
```diff
@@ -70,7 +71,10 @@ function setupLogging(client) {
     });
 
     client.Log.enable();
-    client.Log.entryAdded(entry => process.stderr.write("CDP: " + JSON.stringify(entry["entry"]) + "\n"));
+    client.Log.entryAdded(entry => {
+        logEntries.push(entry["entry"]);
```
Comment (martinpitt) [inline test/common/cdp-driver.js]: Is this related to the unit tests? I thought they use `console.log()`. The CDP `Log` is for messages that Chromium itself generates, not from the JS client script.

File: test/common/cdp.py [hunk 11]
```diff
@@ -170,7 +170,7 @@ def start(self):
 
             # sandboxing does not work in Docker container
             self._browser = subprocess.Popen(
-                argv + ["--disable-gpu", "--no-sandbox", "--remote-debugging-port=%i" % cdp_port, "about:blank"],
+                argv + ["--disable-gpu", "--no-sandbox", "--disable-web-security", "--remote-debugging-port=%i" % cdp_port, "about:blank"],
```
Comment (martinpitt) [inline test/common/cdp.py]: This disables Same-Origin Policy, doesn't it? This will hide actual bugs, so if the unit tests need this for some reason, it should be a (non-default) option of the `CDP` class.

File: test/common/cdp-driver.js (hunk 10 above)
Comment (martinpitt) [reply test/common/cdp-driver.js]: Nevermind, supposedly this is for picking up Resource Errors.

**Comment by martinpitt** [comment]
I agree that the polling and multiplexing between Log and console are ugly. Some random ideas:

 * Re-revert commit 024818d2043 (`test: Revert redirecting the cdp driver's stderr`) and make it an option in the `CDP` constructor. Then integration tests work as before, and unit tests can set the option to read Console (`> ...`) and Log (`CDP: ...`) messages in order from stderr, without polling.

   Note that the `CDP:` prefix for Log isn't magic in any way. I didn't want to use the same prefix as for console (`>`) just to be able to tell them apart. I think this can be dealt with by the regexp matches (accept either prefix).

 * Outputting as JSON instead and always writing to stderr would be okay for me as well, but you then need a separate thread in `class CDP` to read and translate them. I don't like this much TBH, as it could cause messages to be printed out of order, which makes tests harder to debug.

 * Introduce a `CDP.wait_log()` method which doesn't immediately return with the current set of messages (like `get_*()`), but just sets up a promise that gets resolved by both `Log.entryAdded` and `Runtime.consoleAPICalled`. `CDP.command()` is already structured so that it blocks for the promise to get resolved. The resolving can then contain the actual message, so that on the `tap-cdp` side you would just need to call this in a simple loop.

**Comment by larskarlitski** [comment]
Changed as discussed offline with @martinpitt: I've added a cdp.read_log() function which blocks until log messages arrive. This makes `tap-cdp` much easier to read.

I've also removed `--disable-web-security` that I had left in accidentally and made `tap-cdp` not fail anymore when one of the tests fails. This seems counter-intuitive for me, but it consistent with the previous behavior.

I've also added a fix for a hanging test in `src/base1/test-framed-cache.js`. @stefwalter, could you please have a look at that? I'm pretty sure that we can remove the whole `postMessage` thing and just call test.done() when we received all values in the consumer, but I'm not familiar with the semantics of the cache.  

Still WIP because some other tests are still hanging.

**Comment by larskarlitski** [comment]
Update to print messages with `encode('utf-8')`.

Still not happy about the duplicated TAP output. Otherwise tests pass locally now, so I'm removing the WIP.

**Comment by larskarlitski** [comment]
Update to not inject `sizzle.js` and `test_functions.js` when running the unit tests, because the paths are wrong during `make distcheck`. Thanks for helping figure this out, @martinpitt!

File: test/common/cdp.py [hunk 12]
```diff
@@ -226,3 +228,25 @@ def set_frame(self, frame):
...
+
+        if self.valid:
+            # needs to be wrapped in Promise
+            messages = self.command("new Promise((resolve, reject) => resolve(messages))")
+            return map(lambda m: "%s: %s" % tuple(m), messages)
+        return []
+
+    def read_log(self):
+        """Returns an iterator that produces log messages one by one. Blocks if
+        there are no new messages right now."""
```
Comment (martinpitt) [inline test/common/cdp.py]: Short description of a docstring needs to be a single line. I. e. put a paragraph between the sentences.

File: test/common/cdp-driver.js [hunk 13]
```diff
@@ -59,20 +59,56 @@ function success(result) {
...
 
+/**
+ * Returns a promise that resolves when log messages are available. If there
+ * are already some unreported ones in the global messages variable, resolves
+ * immediately.
+ *
+ * Only one such promise can be active at a given time. Once the promise is
+ * resolved, this function can be called again to wait for further messages.
+ */
+function waitLog() {
```
Comment (martinpitt) [inline test/common/cdp-driver.js]: Maybe add an assertion that `logPromiseResolver` is undefined? (For robustness and easier debugging)

File: test/common/cdp.py [hunk 14]
```diff
@@ -226,3 +228,25 @@ def set_frame(self, frame):
...
+
+        if not self.valid:
+            yield []
+            return
+
+        while True:
+            messages = self.command("waitLog()")
+            for m in messages:
+                yield m
```
Comment (martinpitt) [inline test/common/cdp.py]: It feels a bit odd that this is an infinite loop. But the caller exits it at an appropriate time, and I don't have a better idea about it either, so let's go with this for now.

File: test/common/tap-cdp [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+
+import argparse
+import cdp
+import re
+import subprocess
+import sys
+import time
+
+tap_line_re = re.compile(r'^(ok [0-9]+|not ok [0-9]+|bail out!|[0-9]+\.\.[0-9]+|# )', re.IGNORECASE)
+console_line_re = re.compile(r'^([a-z]+): (.*)$')
```
Comment (martinpitt) [inline test/common/tap-cdp]: This is unused.

File: test/common/tap-cdp [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+cdp = cdp.CDP("C.utf8", inject_helpers=False)
+cdp.invoke("Page.navigate", url=address + '/' + opts.test)
+
+ignore_resource_errors = False
+
+for t, message in cdp.read_log():
+
+    # fail on browser level errors
+    if t == 'cdp':
+        if message['level'] != "error" or (ignore_resource_errors and "Failed to load resource" in message["text"]):
```
Comment (martinpitt) [inline test/common/tap-cdp]: It took me a while to mentally invert the condition to verify it, and I'm still not sure about it. Maybe actually invert it and do the abort in the `then` branch, to make this a bit easier to read?

I was thinking it would still be useful to print the message to stderr so that it appears in the log. However, think they actually do already, as we don't redirect cdp-driver.js's  stderr. But doesn't that mean that `make check` would *also* show all the `tap-*` console messages, as they also get written to stderr? It doesn't hurt much, but it might look a bit untidy. Indeed I see a lot of lines like
```
PASS: dist/kubernetes/scripts/test-volumes.html 69 - nfs volume validate: no errors when valid
> log: # module: 
> log: # test: ip_prefix_from_text
> log: ok 1
> log: ok 2
```
in `make check` output.

That said, fine for me to clean this up in a follow-up.

File: test/common/tap-cdp [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+            continue
+        sys.exit(1)
```
Comment (martinpitt) [inline test/common/tap-cdp]: Using `sys.exit` to  fail the test is unclean, as it will skip the `server.terminate` and `cdp.kill`. Maybe go back to what you had before with a `success` flag and use `break` instead?

File: test/common/cdp-driver.js (hunk 13 above)
Comment (larskarlitski) [reply test/common/cdp-driver.js]: Yep!

File: test/common/cdp.py (hunk 12 above)
Comment (larskarlitski) [reply test/common/cdp.py]: +1

File: test/common/cdp.py (hunk 14 above)
Comment (larskarlitski) [reply test/common/cdp.py]: Agreed, I couldn't think of a better way of doing this...

File: test/common/tap-cdp (hunk 15 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Thanks!

File: test/common/tap-cdp (hunk 16 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Yeah that's what I meant with "duplicate output" in an earlier comment. Would be cool if we could address this separately.

Comment (larskarlitski) [reply test/common/tap-cdp]: Oh, and I've cleaned up the condition.

File: test/common/tap-cdp (hunk 17 above)
Comment (larskarlitski) [reply test/common/tap-cdp]: Indeed! Thanks

**Comment by larskarlitski** [comment]
> Also, please add a Closes # to the main commit and capitalize the commit messages, so that this becomes button-mergeable.

Done.

Thanks for the review. I think I've addressed all your points in the latest push.

File: test/common/tap-cdp [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+
+for t, message in cdp.read_log():
+
+    # fail on browser level errors
+    if t == 'cdp':
+        if message['level'] == "error":
+            if ignore_resource_errors and "Failed to load resource" in message["text"]:
+                continue
+            success = False
+            break
```
Comment (martinpitt) [inline test/common/tap-cdp]: Much simpler to understand now, even though it's a bit more verbose. Thanks!
